docs/source/integrations/building-integrations.md

```diff
@@ -1,54 +1,289 @@
 
-<!-- TODO: fix link to point to main once version-4 is merged  -->
-> See the [`expressMiddleware` function](https://github.com/apollographql/apollo-server/blob/36482f5eb56a0421c1eb47e3ebf0e60e033573ab/packages/server/src/express/index.ts) for an example of integrating [Express](https://github.com/expressjs/express) with Apollo Server.
+## Overview
 
-Your integration should accept an `ApolloServer` instance _after_ that instance has called the `server.start()` method:
+This guide demonstrates how to go about building a framework integration for
+Apollo Server at a conceptual level. Below, we will explain the concepts using
+code snippets from our [Express integration](FIXME: link) for illustration.
+
+### Main function signature
+
+Let's start by looking at the main function signature. This snippet uses
+[function
+_overloading_](https://www.typescriptlang.org/docs/handbook/2/functions.html#function-overloads)
+to provide the strongest possible types for the `ApolloServer` instance and the
+user's `context` function. The first two `expressMiddleware` definitions are the
+permitted signatures, while the third is the actual implementation (omitted for
+now).
+
+```ts
+interface ExpressMiddlewareOptions<TContext extends BaseContext> {
+  context?: ContextFunction<[ExpressContextFunctionArgument], TContext>;
+}
+
+export function expressMiddleware(
+  server: ApolloServer<BaseContext>,
+  options?: ExpressMiddlewareOptions<BaseContext>,
+): express.RequestHandler;
+export function expressMiddleware<TContext extends BaseContext>(
+  server: ApolloServer<TContext>,
+  options: WithRequired<ExpressMiddlewareOptions<TContext>, 'context'>,
+): express.RequestHandler;
+export function expressMiddleware<TContext extends BaseContext>(
+  server: ApolloServer<TContext>,
+  options?: ExpressMiddlewareOptions<TContext>,
+): express.RequestHandler {
+  // ...
+}
+```
+
+In the first signature, if `options` is not provided then there is no
+user-provided `context` function to call. The resulting `context` object will
+just be a `BaseContext` (or `{}`). As such, the expected type of the first
+argument is `ApolloServer<BaseContext>`.
+
+The second signature _requires_ that `options` be provided with a `context`
+property, meaning that the type of the `context` object that Apollo Server
+expects is now the same as the return type of the user-provided `context`
+function. We use `TContext` throughout Apollo Server code to represent the
+generic type of the GraphQL context object. We can see that the same `TContext`
+generic is shared by the `ApolloServer` instance as well as the user-provided
+`context` function. This ensures that users type their server and their context
+function correctly.
+
+### Ensure successful startup
+
+For standard integrations, users should await `server.start()` before passing
+their server instance to your integration. This ensures that the server starts
+correctly and that any errors encountered at startup are handled by the user.
```

Comment:
@glasser I should maybe remove this, or just have little opinion about it but explain the approach we took.

Felt funny trying to say "we did this so the handler could be a sync function and you should too".